cranelift/wasm/src/code_translator.rs

```diff
@@ -2479,155 +2479,62 @@ pub fn translate_operator<FE: FuncEnvironment + ?Sized>(
             todo!("not yet implemented: segment.new")
         }
         Operator::SegmentFree { memarg } => {
-            let _ = memarg;
-            todo!("not yet implemented: segment.free")
-        }
-        Operator::SegmentStackNew { memarg } => {
-            /*
-            we build something along the lines of:
-            assert(size is multiple of 16)
-
-            tagged_pointer = irg(memarg)
-            cur_pointer = tagged_pointer
+            // This instruction iterates over the base_ptr (for size bytes) and
+            // tags every 16 bytes of memory with a special free tag.
 
-            loop_start:
-            if (size == 0) goto end
-            stg(cur_pointer, tagged_pointer)
-            size -= 16
-            cur_pointer += 16
-            goto loop_start
+            // Parameters (saved in state):
+            // base_ptr:  pointer to the memory that should be freed by tagging it
+            // size:      size of the memory (in bytes) that should be freed
 
-            end:
-            return tagged_pointer
-             */
+            let special_free_tag: i64 = 0b0000;
 
             let size = state.pop1();
 
-            // First make sure that size is aligned to the next 16 bytes
-            let is_aligned = builder.ins().band_imm(size, 15);
-            builder
-                .ins()
-                .trapnz(is_aligned, ir::TrapCode::HeapMisaligned);
-
-            let (_, base) = unwrap_or_return_unreachable_state!(
+            let (_, base_ptr) = unwrap_or_return_unreachable_state!(
                 state,
-                prepare_addr(memarg, 32, builder, state, environ)?
+                prepare_addr_impl(memarg, 32, builder, state, environ, true)?
             );
 
-            // TODO: We also need to write optimizations ourselves that are able to merge stg into st2g
-            let tagged_pointer = builder.ins().arm64_irg(base);
-
-            let inner_block = block_with_params(builder, [ValType::I32, ValType::I64], environ)?;
-            let next_inner_block =
-                block_with_params(builder, [ValType::I32, ValType::I64], environ)?;
-            let next_block = block_with_params(builder, [ValType::I64], environ)?;
-
-            builder.ins().jump(inner_block, &[size, tagged_pointer]);
+            // remove existing tag in base_ptr
+            let base_ptr = builder.ins().band_imm(base_ptr, 0x0FFF_FFFF_FFFF_FFFF);
```

Comment:
MTE tag is bits 56-59 (sorry -- my fault!).

```suggestion
            let base_ptr = builder.ins().band_imm(base_ptr, 0xF0FF_FFFF_FFFF_FFFF);
```